subscription_oca/models/sale_subscription.py

```diff
@@ -0,0 +1,807 @@
+                field_name = "standard_price"
+                product_currency = product.cost_currency_id
+            elif (
+                pricelist_item.base == "pricelist" and pricelist_item.base_pricelist_id
+            ):
+                field_name = "price"
+                product = product.with_context(
+                    pricelist=pricelist_item.base_pricelist_id.id
+                )
+                product_currency = pricelist_item.base_pricelist_id.currency_id
+            currency_id = pricelist_item.pricelist_id.currency_id
+
+        if not currency_id:
+            currency_id = product_currency
+            cur_factor = 1.0
+        else:
+            if currency_id.id == product_currency.id:
+                cur_factor = 1.0
+            else:
+                cur_factor = currency_id._get_conversion_rate(
+                    product_currency,
+                    currency_id,
+                    self.company_id or self.env.company,
+                    fields.Date.today(),
+                )
+
+        product_uom = self.env.context.get("uom") or product.uom_id.id
+        if uom and uom.id != product_uom:
+            # the unit price is in a different uom
+            uom_factor = uom._compute_price(1.0, product.uom_id)
+        else:
+            uom_factor = 1.0
+
+        return product[field_name] * uom_factor * cur_factor, currency_id
+
+    def _get_display_price(self, product):
+        if self.sale_subscription_id.pricelist_id.discount_policy == "with_discount":
+            return product.with_context(
+                pricelist=self.sale_subscription_id.pricelist_id.id,
+                uom=self.product_id.uom_id.id,
+            ).price
+
+        final_price, rule_id = self.sale_subscription_id.pricelist_id.with_context(
+            partner_id=self.sale_subscription_id.partner_id.id,
+            date=fields.Datetime.now(),
+            uom=self.product_id.uom_id.id,
+        ).get_product_price_rule(
+            product or self.product_id,
+            self.product_uom_qty or 1.0,
+            self.sale_subscription_id.partner_id,
+        )
+        base_price, currency = self.with_context(
+            partner_id=self.sale_subscription_id.partner_id.id,
+            date=fields.Datetime.now(),
+            uom=self.product_id.uom_id.id,
+        )._get_real_price_currency(
+            product, rule_id, self.product_uom_qty, self.product_id.uom_id
+        )
+        if currency != self.sale_subscription_id.pricelist_id.currency_id:
+            base_price = currency._convert(
+                base_price,
+                self.sale_subscription_id.pricelist_id.currency_id,
+                self.sale_subscription_id.company_id or self.env.company,
+                fields.Date.today(),
+            )
+        return max(base_price, final_price)
+
+    def _compute_tax_id(self):
+        for line in self:
+            line = line.with_company(line.company_id)
```

Comment:
there's no need to call with_company for each line IMO, no?
If needed, you can group lines by company.